sycl/include/CL/sycl/detail/resource_pool.hpp

```diff
@@ -0,0 +1,274 @@
+      return lhs.Size < rhs;
+    }
+    bool operator()(size_t lhs, FreeEntry const &rhs) const {
+      return lhs < rhs.Size;
+    }
+  };
+
+  /// Extracts a free entry from the pool that fits the size required. If there
+  /// is no suitable entry, new memory will be allocated.
+  ///
+  /// \param Range is the range of the resulting buffer.
+  /// \param ContextImplPtr is the context to allocate memory in.
+  /// \param DataPtr is the pointer to data on the host to initialize the
+  ///        associated memory with. This will only be used if a new entry is
+  ///        allocated.
+  /// \param IsNewEntry will be set to true if the entry was newly allocated in
+  ///        the pool and false if it was found in the existing free entries in
+  ///        the pool. This is not set if it is nullptr.
+  /// \return a shared pointer to the new managed resource.
+  FreeEntry
+  getOrAllocateEntry(const size_t Size,
+                     const std::shared_ptr<context_impl> &ContextImplPtr,
+                     void *DataPtr = nullptr, bool *IsNewEntry = nullptr);
+
+  /// Extracts a free entry from the pool that fits the size required. If there
+  /// is no suitable entry, new memory will be allocated. The memory will be
+  /// initialized with the data given.
+  ///
+  /// \param Size is the size of the free entry to find or allocate.
+  /// \param QueueImplPtr is the queue with the context to allocate memory in.
+  /// \param DataPtr is the pointer to data on the host to initialize the
+  ///        associated memory with.
+  /// \param AvailableEvent will be set to an event that is tied to the
+  ///        initialization of the memory.
+  /// \param IsNewEntry will be set to true if the entry was newly allocated in
+  ///        the pool and false if it was found in the existing free entries in
+  ///        the pool. This is not set if it is nullptr.
+  /// \return a shared pointer to the new managed resource.
+  FreeEntry getOrAllocateEntry(const size_t Size,
+                               const std::shared_ptr<queue_impl> &QueueImplPtr,
+                               void *DataPtr, event *AvailableEvent,
+                               bool *IsNewEntry = nullptr);
+
+  /// Gets the context implementation associtated with a queue implementation.
+  ///
+  /// \param QueueImplPtr is the queue implementation to get the context
+  /// implementation from. \return the context implementation from the queue
+  /// implementation.
+  static const std::shared_ptr<context_impl> &
+  getQueueContextImpl(const std::shared_ptr<queue_impl> &QueueImplPtr);
+
+  using ContextPtr = context_impl *;
+
+public:
+  /// Removes and deallocates all free entries currently in the pool.
+  void clear();
+
+  ResourcePool();
+  ResourcePool(const ResourcePool &) = delete;
+  ~ResourcePool() {
+    clear();
+    assert(MAllocCount == 0 && "Not all resources have returned to the pool.");
+  }
+
+  /// Sets the platform of the resource pool.
```

Comment:
why set and track the platform as opposed to the context?   Is the memory pool per-platform or per-context?  
And, in any case, can't we set that from getOrAllocateEntry  (or possibly getOrAllocateResource) ?

Reply:
Platform was picked over context because all we actually need is the plugin, which the context gets from platform anyway. It could in principle hold a raw pointer (or potentially a `weak_ptr`) to context, which was also the first iteration of it, but there's not much benefit to it.

That said, I think you are right we could set it first time we see it. I will try it out and report back!

Reply:
It seems to work just fine. I have made the change to setting it during `getOrAllocateEntry`.